lib/apply_state_change.js

```diff
@@ -0,0 +1,19 @@
+function applyStateChange(state, change) {
+  const { path, action, value } = change;
+
+  switch (action) {
+    case "set":
+      if (path.length) return state.mergeIn(path, value);
+      else return value;
+    case "add":
+      const list = state.getIn(path).push(value);
+      return state.setIn(path, list);
+    case "remove":
+      return state.removeIn(path);
+    case "replace":
+      if (path.length) return state.setIn(path, value);
```

Comment:
seems like immutable.js does the same thing when you `setIn` on an empty path, returning the value directly. so, `if` might be extra here.